### PR TITLE
wip: github components kit

### PR DESCRIPTION
**Pull request by sidmohanty11**
- Added a new GithubIssuesList Component taking inspiration from discourse ranked list component
- Used markedjs to parse the markdown text given in issue body and created a new MDPreview component which can be reused.
- Likes are based on no. of upvotes / +1s 
- Comments are the no. of comments
- For now, it's showing upto 10 issues
- Shows upto 20 words max and then truncates (of issue body)

rc4_comm_gitint1

#27

### CODE REVIEW

File: app/pages/index.js [hunk 1]
```diff
@@ -81,8 +89,15 @@ export async function getStaticProps({ params }) {
   const topNavItems = await fetchAPI('/top-nav-item');
   const topPosts = await fetchAPI('/discourses');
 
+  const i = await fetch('https://api.github.com/repos/RocketChat/RC4Community/issues', {
```
Comment (debdutdeb) [inline app/pages/index.js:92]: if we're planning on extending github integration, better use octokit than any http client.

Comment (sidmohanty11) [reply app/pages/index.js]: I've used octokit now, can you please check?

Comment (sidmohanty11) [reply app/pages/index.js]: I thought it will add as a new dependency to our application but I think it can be very useful during extension!

**Comment by debdutdeb** [comment]
Forcing workflow run.

**Review by Sing-Li — CHANGES_REQUESTED** [review]
Add a compact mode for the component.   Full page components are NOT THAT useful.   We may as well just create a route for it.   Imagine it is embedded as one of 4 components in a row of the page, restyle the component under this circumstance.

Typically, this should also be the reactive mode display when the width is shrunk (or using mobile browser).

**Comment by sidmohanty11** [comment]
@Sing-Li the way I am imagining it is, 

- we can make another reusable GithubIssueCard component which will be styled like a proper Github Issue and on the home page under "Trending issues" or something like that we can add the top 4 or 5 issues (which can be sorted by most upvoted or comments) 
- then we can use that card component in the route we create (for example, /gh-issues) to show the top 10-15 issues

Please suggest your opinions and I'll start working on it!

File: app/lib/octokit.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
+import { Octokit } from '@octokit/core';
```
Comment (debdutdeb) [inline app/lib/octokit.js:1]: Convert this whole to a github library of sorts.

File: app/pages/index.js [hunk 3]
```diff
@@ -9,6 +9,8 @@ import Growthcounters from '../components/growthcounters';
 import { Container, Col } from 'react-bootstrap';
 import { fetchAPI } from '../lib/api';
 import { withFirebaseAuthUser } from '../components/auth/firebase';
+import GithubIssuesList from '../components/githubissueslist';
+import { octokitClient } from '../lib/octokit'
```
Comment (debdutdeb) [inline app/pages/index.js:13]:
```suggestion
import { getIssues } from '../lib/github'
```

Consider these just suggestions, do it your own way, no need to commit these exact ones.

**Comment by Sing-Li** [comment]
Great ideas @sidmohanty11 !     Really do like the two components idea.   But please do name it consistent with bootstrap culture (I don't know if "card" is the right word in that design language - sounds very Material Design'ish).  Also  please name the two components appropriately so that the "large" one is for 1/3, 1/2, 1 page layouts where the "card?" one is for finer compositions.  

As far as aesthetics and layout go - please confer in DM directly with @demonicirfan  .... he's got some extreme "design chops" and may be able to figma you some workable ideas.    Thanks.

**Review by debdutdeb — CHANGES_REQUESTED** [review]
I think there is a lot of potential for github integration, and for that reason I suggested using octokit from that start. 

Since we're now using octokit, imo it'd be better to not have a `octokit` library (which also confuses people with main `octokit`), but rather a collection of helper functions for grabbing common github resources.

**Comment by debdutdeb** [comment]
@sidmohanty11 if you have any questions or need any clarifications for the changes I requested, feel free to ask :)

**Comment by sidmohanty11** [comment]
@Sing-Li Got it! 

I've already DMed him, as soon as he comes up with an idea I will start the execution part.

**Comment by sidmohanty11** [comment]
@debdutdeb thank you so much for the suggestions! 

Just so I am crystal clear, I am going to make a `lib/github.js` file where all the helper functions will reside like getIssues, getPRs (you can give some other suggestions too) etc? And return data as their purpose states directly by just passing in `params`? and export them all. Did I get it correctly?

For example, 
```javascript
const issues = await getIssues('RocketChat', 'RC4Community'); // will return all issues
const prs = await getPRs('RocketChat', 'RC4Community'); // will return all PRs
```

Thank you

**Comment by Sing-Li** [comment]
@sidmohanty11   perhaps make this into a 3 components  -   Github components kit.   If you add this component - I am sure it will eventually become the most used out of all the Github components 😃 

image

**Comment by sidmohanty11** [comment]
@Sing-Li Amazing suggestion! I will start working

**Comment by debdutdeb** [comment]
@Sing-Li like the idea 💯 

Just a suggestion, maybe we don't change and mix the goal of this pr, and merge it as it is, then move the component kit to a seperate pr?

**Comment by Sing-Li** [comment]
Thx 4 suggestion.  

All the Github related components in one "kit" or PR is best, IMO.  Each component is actually quite trivial at the moment - and they'll use many common APIs.   Otherwise we might end up with too much to manage.

We've been doing 3 or 4 components PRs quite frequently so far.     There will likely be another 30 to 40 to be done yet before we get a reasonable foundation for the project.

**Comment by sidmohanty11** [comment]
- Added new contributorslist component which shows the top 100 contributors,
rc4_comm_contributors1
- Arranged all the github related components inside a separate github folder which can be useful for further expansion (for example addition of other components related to github integration)
- Made some helper functions `getIssues`, `getPRs`, `getContributors` and exported it from `lib/github.js` @debdutdeb please let me know if I need to make some other changes

- [ ] The re-styling of github issue "card" isn't completed yet

Thank you

**Comment by Sing-Li** [comment]
@sidmohanty11  really looks good.   I'll give it a test now.

**Comment by Sing-Li** [comment]
@sidmohanty11   I've tested everything and basically working fine.   I also liked how you've structured the code, making it very easy to migrate to the next step.  

Which is to cache the data (issues, etc) through strapi, the headless CMS, like all the other getStaticProps() fetch all do.

This is important to lighten the load on Github API calls  if the site re-generation frequency is high.

Please use the currently open Leaderboard component PR as an example.  It should be quite easy now that you have your data fetch functions all under a `lib` directory.   

Thanks.

**Comment by sidmohanty11** [comment]
@Sing-Li I've added the cron job which will make fetch requests after an hour and an initial fetch request during initialisation of data(so that we don't wait for data for 1st 1hr). And I've checked that it will be very rare but if at any edge case situation we change the cron request time to 60secs or so..  we might  get an error that says we have exceeded the fetch request quota. So for that I've just made a default issue template otherwise in this way the cron job will update the Issues accordingly. :-
### if maxed out 
rc_comm_ifmaxed
 ### normal fetch
rc_comm_notmaxed

Please give it a test and provide me some of your suggestions

Thank you

**Comment by Sing-Li** [comment]
This is awesome  @sidmohanty11 !   I'll test this tonight or over the weekend.  

In the meantime, please add component kit  documentation (following standard ReactJS component doc format, no the leaderboard component's format)   here:

```
https://github.com/RocketChat/RC4Community/pull/66
```

Thanks.

**Comment by Sing-Li** [comment]
@sidmohanty11  the URL to docs is updated above - https://github.com/RocketChat/RC4Community/tree/master/docs/components

**Comment by sidmohanty11** [comment]
@Sing-Li sorry for the late reply, I went to visit Grandma this weekend. I will finish the work by today. Thank you

**Comment by sidmohanty11** [comment]
@Sing-Li I've added the documentation and have tried to be as much descriptive as I could, still if I've missed something or for any improvements, I would love to get your suggestions

gh-components-kit-docs

**Pull request merged**